### PR TITLE
[snmp] Add test cases for SONiC physical entity MIB feature

### PR DESCRIPTION
**Pull request by Junchao-Mellanox**
### Description of PR

Summary:
Add test cases for SONiC physical entity MIB feature

### Type of change

- [ ] Bug fix
- [ ] Testbed and Framework(new/improvement)
- [x] Test case(new/improvement)

#### What is the motivation for this PR?

There is no regression test case for physical entity MIB before, and we create a few new test cases to cover the existing/new MIB objects.

#### How did you do it?

1. Adjust snmp_facts.py to support new MIB objects
2. Add test case to verify transceiver related entities
3. Add test case to verify PSU related entities
4. Add test case to verify fan related entities
5. Add test case to verify thermal related entities
6. Add test case to verify PSU/fan remove 

#### How did you verify/test it?

Run the new regression test cases 

#### Any platform specific information?

N/A

#### Supported testbed topology if it's a new test case?

N/A

### CODE REVIEW

**Comment by Junchao-Mellanox** [comment]
Depends on:
https://github.com/Azure/sonic-platform-common/pull/134
https://github.com/Azure/sonic-platform-daemons/pull/102
https://github.com/Azure/sonic-buildimage/pull/5645
https://github.com/Azure/sonic-snmpagent/pull/168

**Comment by Junchao-Mellanox** [comment]
retest vsimage please

**Comment by wangxin** [comment]
retest this please

**Comment by lguohan** [comment]
can you please add the test to kvmtest.sh?

**Comment by Junchao-Mellanox** [comment]
> kvmtest.sh

Done.

**Comment by Junchao-Mellanox** [comment]
retest vsimage please

**Comment by Junchao-Mellanox** [comment]
retest vsimage please

**Comment by liat-grozovik** [comment]
retest vsimage please

**Comment by liat-grozovik** [comment]
@Junchao-Mellanox kindly reminder that this feature will go only to master towards 202012. 
As of that we need a way to run snmp test on 201911 which has no such capability. Should it work or need to find a solution to detect if the version expected to support this extended mib?

**Comment by Junchao-Mellanox** [comment]
Hi @liat-grozovik , we can do this in sonic-tool. In sonic-tool, we can disable cases according to the image branch name. So once we merge this test case to our local repo, we need add the case to sonic-tool and disable it on 201911 branch in snmp.cases file.

**Comment by Junchao-Mellanox** [comment]
retest vsimage please

**Comment by keboliu** [comment]
retest vsimage please

**Comment by keboliu** [comment]
retest this please

**Comment by Junchao-Mellanox** [comment]
retest vsimage please

**Comment by Junchao-Mellanox** [comment]
retest vsimage please

**Comment by Junchao-Mellanox** [comment]
retest vsimage please

**Comment by Junchao-Mellanox** [comment]
retest vsimage please

**Comment by Junchao-Mellanox** [comment]
retest vsimage please

**Review by keboliu — APPROVED** [review]

File: tests/snmp/test_snmp_phy_entity.py [hunk 1]
```diff
@@ -0,0 +1,583 @@
...
+                        SENSOR_TYPE_PORT_RX_POWER + 2,
+                        SENSOR_TYPE_PORT_RX_POWER + 3,
+                        SENSOR_TYPE_PORT_RX_POWER + 4,
+                        SENSOR_TYPE_PORT_TX_BIAS + 1,
+                        SENSOR_TYPE_PORT_TX_BIAS + 2,
+                        SENSOR_TYPE_PORT_TX_BIAS + 3,
+                        SENSOR_TYPE_PORT_TX_BIAS + 4,
+                        SENSOR_TYPE_VOLTAGE]
+
+# Redis Constants
```
Comment (qiluo-msft) [inline tests/snmp/test_snmp_phy_entity.py:91]: >Redis [](start = 2, length = 5)

Suggest remove `Redis`

File: tests/snmp/test_snmp_phy_entity.py [hunk 2]
```diff
@@ -0,0 +1,583 @@
...
+        assert fan_snmp_fact['entPhysName'] == name
+        assert fan_snmp_fact['entPhysHwVer'] == ''
+        assert fan_snmp_fact['entPhysFwVer'] == ''
+        assert fan_snmp_fact['entPhysSwVer'] == ''
+        assert fan_snmp_fact['entPhysSerialNum'] == '' if is_null_str(fan_info['serial']) else fan_info[
+            'serial']
+        assert fan_snmp_fact['entPhysMfgName'] == ''
+        assert fan_snmp_fact['entPhysModelName'] == '' if is_null_str(fan_info['model']) else fan_info['model']
+        assert fan_snmp_fact['entPhysIsFRU'] == REPLACEABLE if fan_info[
+                                                                   'is_replaceable'] == 'True' else NOT_REPLACEABLE
```
Comment (qiluo-msft) [inline tests/snmp/test_snmp_phy_entity.py:209]: >'is_replaceable' [](start = 67, length = 16)

Line breaking at a weird position.

File: tests/snmp/test_snmp_phy_entity.py [hunk 3]
```diff
@@ -0,0 +1,583 @@
...
+def test_psu_info(duthost, snmp_physical_entity_info):
+    """
+    Verify PSU information in physical entity mib with redis database
+    :param duthost: DUT host object
+    :param snmp_physical_entity_info: Physical entity information from snmp fact
+    :return:
+    """
+    keys = redis_get_keys(duthost, STATE_DB, PSU_KEY_TEMPLATE.format('*'))
+    if not keys:
+        pytest.skip('PSU information not exists in DB, skipping this test')
```
Comment (qiluo-msft) [inline tests/snmp/test_snmp_phy_entity.py:238]: >not exists [](start = 37, length = 10)

does not exist

Comment (qiluo-msft) [inline tests/snmp/test_snmp_phy_entity.py:238]: >pytest.skip [](start = 8, length = 11)

There are many `pytest.skip()`. Seems are a valid failure case.
Do you skip just for backward-compatible? If yes, I think this is not a good method.

File: tests/snmp/test_snmp_phy_entity.py [hunk 4]
```diff
@@ -0,0 +1,583 @@
...
+            assert expect_oid not in snmp_physical_entity_info
+            continue
+
+        assert expect_oid in snmp_physical_entity_info, 'Cannot find PSU sensor {} in physical entity mib'.format(field)
+        sensor_snmp_fact = snmp_physical_entity_info[expect_oid]
+        assert sensor_snmp_fact['entPhysDescr'] == '{} for {}'.format(sensor_tuple[0], psu_name)
+        assert sensor_snmp_fact['entPhysContainedIn'] == psu_oid
+        assert sensor_snmp_fact['entPhysClass'] == PHYSICAL_CLASS_SENSOR
+        assert sensor_snmp_fact['entPhyParentRelPos'] == sensor_tuple[1]
+        assert sensor_snmp_fact['entPhysName'] == '{} for {}'.format(sensor_tuple[0], psu_name)
```
Comment (qiluo-msft) [inline tests/snmp/test_snmp_phy_entity.py:292]: >'{} for {}'.format(sensor_tuple[0], psu_name) [](start = 50, length = 45)

Appear twice, use a variable?

File: tests/snmp/test_snmp_phy_entity.py [hunk 5]
```diff
@@ -0,0 +1,583 @@
...
+        pytest.skip('psu_controller is None')
+    psu_status = psu_controller.get_psu_status()
+    if len(psu_status) < 2:
+        pytest.skip('At least 2 PSUs required for rest of the testing in this case')
+
+    # turn on all PSU
+    for item in psu_status:
+        if not item['psu_on']:
+            psu_controller.turn_on_psu(item["psu_id"])
+            time.sleep(5)
```
Comment (qiluo-msft) [inline tests/snmp/test_snmp_phy_entity.py:421]: >time.sleep(5) [](start = 12, length = 13)

What if no sleep?

File: tests/snmp/test_snmp_phy_entity.py [hunk 6]
```diff
@@ -0,0 +1,583 @@
...
+
+def redis_get_keys(duthost, db_id, pattern):
+    """
+    Get all keys for a given pattern in given redis database
+    :param duthost: DUT host object
+    :param db_id: ID of redis database
+    :param pattern: Redis key pattern
+    :return: A list of key name in string
+    """
+    cmd = 'redis-cli --raw -n {} KEYS \"{}\"'.format(db_id, pattern)
```
Comment (qiluo-msft) [inline tests/snmp/test_snmp_phy_entity.py:522]: >redis-cli [](start = 11, length = 9)

`redis-cli` will not aware of multi-DB or multi-namespace. Use `sonic-db-cli`

**Review by qiluo-msft — CHANGES_REQUESTED** [review]
As comments

File: tests/snmp/test_snmp_phy_entity.py (hunk 3 above)
Comment (Junchao-Mellanox) [reply tests/snmp/test_snmp_phy_entity.py:238]: I skip this for vsimage test. I found this testcase will be ran in vsimage, and there is no PSU information in DB.

File: tests/snmp/test_snmp_phy_entity.py (hunk 5 above)
Comment (Junchao-Mellanox) [reply tests/snmp/test_snmp_phy_entity.py:421]: The PSU may take a few seconds to power on, so I add a sleep here, this code is similar with some test cases in platform tests.

Comment (qiluo-msft) [reply tests/snmp/test_snmp_phy_entity.py:420]: Better to sleep only once after all PSU turned on?

---
In reply to: [532159395](https://github.com/Azure/sonic-mgmt/pull/2379#discussion_r532159395) [](ancestors = 532159395)

File: tests/snmp/test_snmp_phy_entity.py (hunk 3 above)
Comment (qiluo-msft) [reply tests/snmp/test_snmp_phy_entity.py]: @wangxin What is best practice to skip test case in vsimage test?

---
In reply to: [532159260](https://github.com/Azure/sonic-mgmt/pull/2379#discussion_r532159260) [](ancestors = 532159260)

Comment (wangxin) [reply tests/snmp/test_snmp_phy_entity.py]: You can mark the test case like below to specify that this test case is only applicable for physical testbed, not for vsimage.
```
@pytest.mark.device_type('physical')
def test_xxx
```

**Comment by Junchao-Mellanox** [comment]
Hi @wangxin , it seems the test was not skipped in vsimage after adding the pytest mark. I see following error in the test result:
```
11:08:57      @pytest.mark.device_type('physical')
11:08:57      def test_fan_info(duthost, snmp_physical_entity_info):
11:08:57          """
11:08:57          Verify fan information in physical entity mib with redis database
11:08:57          :param duthost: DUT host object
11:08:57          :param snmp_physical_entity_info: Physical entity information from snmp fact
11:08:57          :return:
11:08:57          """
11:08:57          keys = redis_get_keys(duthost, STATE_DB, FAN_KEY_TEMPLATE.format('*'))
11:08:57  >       assert keys, 'Fan information does not exist in DB'
11:08:57  E       AssertionError: Fan information does not exist in DB
```

**Comment by wangxin** [comment]
If all the test cases in this test script need physical testbed, then it should not be added to the list in `kvmtest.sh`.

**Review by qiluo-msft — APPROVED** [review]

**Review by wangxin — APPROVED** [review]

**Pull request merged**